python/tank/platform/engine.py

```diff
@@ -959,33 +1001,59 @@ def __create_main_thread_invoker(self):
             if QtGui and QtCore:
                 class Invoker(QtCore.QObject):
+                    """
+                    Invoker class - implements a mechanism to execute a function with arbitrary
+                    args in the main thread.
+                    """
                     def __init__(self):
+                        """
+                        Construction
+                        """
                         QtCore.QObject.__init__(self)
+                        self._lock = threading.Lock()
+                        self._fn = None
                         self._res = None
-                        
+
                     def invoke(self, fn, *args, **kwargs):
-                        self._fn = lambda: fn(*args, **kwargs) 
-                        self._res = None
-                        
-                        QtCore.QMetaObject.invokeMethod(self, "_do_invoke", QtCore.Qt.BlockingQueuedConnection)
-                        
-                        return self._res
-                
+                        """
+                        Invoke the specified function with the specified args in the main thread
+
+                        :param fn:          The function to execute in the main thread
+                        :param *args:       Args for the function
+                        :param **kwargs:    Named arguments for the function
+                        :returns:           The result returned by the function
+                        """
+                        # acquire lock to ensure that the function and result are not overwritten 
+                        # by syncrounous calls to this method from different threads
+                        self._lock.acquire()
+                        try:
+                            self._fn = lambda: fn(*args, **kwargs)
+                            self._res = None
+
+                            # invoke the internal _do_invoke method that will actually run the function.  Note that
+                            # we are unable to pass/return arguments through invokeMethod as this isn't properly
+                            # supported by PySide.
+                            QtCore.QMetaObject.invokeMethod(self, "_do_invoke", QtCore.Qt.BlockingQueuedConnection)
+
+                            return self._res
+                        finally:
+                            self._lock.release()
+
                     @qt.QtCore.Slot()
                     def _do_invoke(self):
                         """
-                        Execute function and return result
+                        Execute the function
                         """
                         self._res = self._fn()
 
-                # Make sure that the invoker is for the main thread only.
+                # Make sure that the invoker exists in the main thread:
                 invoker = Invoker()
                 invoker.moveToThread(QtCore.QThread.currentThread())
```

Comment:
I think this line is wrong - shouldn't it be moving to the main thread (QtGui.QApplication.instance().thread())?